src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,284 @@
+                    };
+                    this.dataStore.Insert(this.data);
+                }
+
+                this.blockReceivedQueue = new AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>>(this.OnEnqueueAsync);
+
+                // Subscribe to events.
+                this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+                {
+                    while (this.blockReceivedQueue.Count > 100)
+                    {
+                        this.logger.LogWarning("Address indexing is slowing down the consensus.");
+                        Thread.Sleep(5000);
+                    }
+
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(true, blockConnectedData.ConnectedBlock));
+                });
+
+                this.blockDisconnectedSubscription = this.signals.Subscribe<BlockDisconnected>(blockDisconnectedData =>
+                {
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(false, blockDisconnectedData.DisconnectedBlock));
+                });
+
+                this.nodeStats.RegisterStats(this.AddComponentStats, StatsType.Component, 400);
+            }
+        }
+
+        private Task OnEnqueueAsync(KeyValuePair<bool, ChainedHeaderBlock> item, CancellationToken cancellationtoken)
+        {
+            try
+            {
+                this.ProcessBlockAddedOrRemoved(item);
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex.ToString());
+                throw;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AddComponentStats(StringBuilder benchLog)
+        {
+            benchLog.AppendLine();
+            benchLog.AppendLine("======AddressIndexer======");
+
+            benchLog.AppendLine($"Unprocessed blocks: {this.blockReceivedQueue.Count}");
+        }
+
+        /// <summary>Processes block that was added or removed from consensus chain.</summary>
+        private void ProcessBlockAddedOrRemoved(KeyValuePair<bool, ChainedHeaderBlock> item)
+        {
+            // Make sure it's on top of the tip.
+            bool blockAdded = item.Key;
+
+            if ((blockAdded && item.Value.ChainedHeader.Header.HashPrevBlock.ToString() != this.data.TipHash) ||
+                (!blockAdded && item.Value.ChainedHeader.HashBlock.ToString() != this.data.TipHash))
+            {
+                const string message = "TransactionIndexer is in inconsistent state. This can happen if you've enabled txindex on an already synced or partially synced node. " +
+                                       "Remove everything from the data folder and run the node with -txindex=true.";
+
+                this.logger.LogCritical(message);
+                this.logger.LogTrace("(-)[INCONSISTENT_STATE]");
+                throw new Exception(message);
```

Comment:
There is a specific exception we can throw I beleive.

Reply:
Why? It's not supposed to be handled automatically